// microsoft-cognitiveservices-speech-sdk/distrib/es2015/src/common.browser/ProxyInfo.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
import { PropertyId } from "../sdk/Exports";
export class ProxyInfo {
    constructor(proxyHostName, proxyPort, proxyUserName, proxyPassword) {
        this.privProxyHostName = proxyHostName;
        this.privProxyPort = proxyPort;
        this.privProxyUserName = proxyUserName;
        this.privProxyPassword = proxyPassword;
    }
    static fromRecognizerConfig(config) {
        return new ProxyInfo(config.parameters.getProperty(PropertyId.SpeechServiceConnection_ProxyHostName), parseInt(config.parameters.getProperty(PropertyId.SpeechServiceConnection_ProxyPort), 10), config.parameters.getProperty(PropertyId.SpeechServiceConnection_ProxyUserName), config.parameters.getProperty(PropertyId.SpeechServiceConnection_ProxyPassword));
    }
    get HostName() {
        return this.privProxyHostName;
    }
    get Port() {
        return this.privProxyPort;
    }
    get UserName() {
        return this.privProxyUserName;
    }
    get Password() {
        return this.privProxyPassword;
    }
}
